### PR TITLE
index comments

### PR DESCRIPTION
**Pull request by SebastianSchmidl**
see description of Assignment 2

### CODE REVIEW

File: cse/lang/__init__.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    __tokenizer = None
+    __stemmer = None
+    __lemmatizer = None
+
+
+    def __init__(self):
+        self.__tokenizer = NltkTokenizer()
+
+
+    def useNltkTokenizer(self):
```
Comment (Benedikt1992) [inline cse/lang/__init__.py]: It's clear that we need a tokenizer. But if a `useNltkTokenizer` method exists it should be optional (additionally it's already initialized in the `__init__` method)

Comment (SebastianSchmidl) [reply cse/lang/__init__.py]: So would you suggest to silently use the `NltkTokenizer` as default (in the `__init__` method) if there was none specified? Or do you think, it's better to let the user decide and if none specified raise an exception (like in [line 72](https://github.com/CodeLionX/CommentSearchEngine/pull/19/files#diff-9e24383affb7ae83521fe57adb80f94fR72))?

Comment (Benedikt1992) [reply cse/lang/__init__.py]: If we have a method `useTokenizer` we could define the NltkTokenizer as default. For now I would raise an exception.

**Review by Benedikt1992 — APPROVED** [review]

File: cse/CommentReader.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    def readData(self):
+        first = True
+        second = True
+
+        articleUrl = ""
+        articleId = ""
+        comments = {}
+
+        for row in self.__reader:
```
Comment (Benedikt1992) [inline cse/CommentReader.py:38]: You read every line but only process somethin at the second row?

File: cse/lang/NltkStemmer.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        self.__stemmerType = stemmerType
+
+
+    @staticmethod
+    def porter():
+        from nltk.stem.porter import PorterStemmer
+        return PorterStemmer()
+
+    @staticmethod
+    def porter2():
```
Comment (Benedikt1992) [inline cse/lang/NltkStemmer.py:17]: As far as I know you have to download the language set in order to use it. Do we need to add something in the setup process of our developing box?

File: cse/lang/NltkStopwordFilter.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    def __init__(self, stopwordList):
+        # lookup is faster in a set than a list
+        self.__stopwordlist = set(stopwordList)
+
+
+    @staticmethod
+    def english():
+        from nltk.corpus import stopwords
+        return stopwords.words("english")
```
Comment (Benedikt1992) [inline cse/lang/NltkStopwordFilter.py:16]: Same here. Do we need to download the list for english stopwords?

File: cse/lang/NltkTokenizer.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    def tokenize(self, text):
+        # tokenize into sentences and sentences into lower case words
+        tokens = [word.lower() for sent in sent_tokenize(text) for word in word_tokenize(sent)]
+        # filter out punctuation
+        tokens = filter(lambda word: word not in string.punctuation, tokens)
+        # replace "n't" with "not"
+        tokens = map(lambda word: word if word != "n't" else "not", tokens)
+        tokens = [re.sub(r'^-|-$|\'', '', word) for word in tokens]
+        tokens = filter(lambda word: word, tokens)
```
Comment (Benedikt1992) [inline cse/lang/NltkTokenizer.py:21]: What is this statement supposed to do?

File: cse/indexing/MainPostingListIndex.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    def estimatedSize(self):
+        return self.__sizeof__()
+
+
+    def mergeInDeltaIndex(self, dIndex):
+        self.__postingLists.seek(0)
+        _, tempFilePath = mkstemp()
+        visited = set()
+        with open(tempFilePath, 'w', newline='', encoding="utf-8") as tempFile:
```
Comment (Benedikt1992) [inline cse/indexing/MainPostingListIndex.py:68]: In L66 you already got an filehandle (`_` - the name could be more significant).  Why open the file again?

File: cse/indexing/MainPostingListIndex.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                if i in dIndex:
+                    postingList = self.__decodePlLine(plLine)
+                    postingList = postingList + dIndex[i]
+                    postingList.sort()
+                    tempFile.write(self.__encodePlLine(postingList))
+                else:
+                    tempFile.write(plLine)
+                visited.add(i)
+            for pointer in dIndex:
+                if pointer not in visited:
```
Comment (Benedikt1992) [inline cse/indexing/MainPostingListIndex.py:79]: In this statement you assume, that the `dIndex`-dict is ordered. This is wrong.

https://stackoverflow.com/questions/9001509/how-can-i-sort-a-dictionary-by-key

File: cse/indexing/createIndex.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import os
+from cse.indexing.FileIndex import FileIndex
```
Comment (Benedikt1992) [inline cse/indexing/createIndex.py:2]: Does not exist?

File: cse/indexing/InvertedIndexReader.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    def close(self):
+        self.__dictionary.close()
+        self.__mIndex.close()
+
+
+    def retrieve(self, term):
+        if term in self.__dictionary:
+            pointer = self.__dictionary[term]
+            return self.__mIndex[pointer]
```
Comment (Benedikt1992) [inline cse/indexing/InvertedIndexReader.py:28]: Out of curiosity: How is it implemented that you can call `[]` on the `MainPostingListIndex`-Class?

File: cse/lang/NltkTokenizer.py (hunk 5 above)
Comment (SebastianSchmidl) [reply cse/lang/NltkTokenizer.py:21]: nothing 😄 it's a remnant

File: cse/lang/NltkStemmer.py (hunk 3 above)
Comment (SebastianSchmidl) [reply cse/lang/NltkStemmer.py:17]: Yeah, you're right. I added it to the `installDependencies` script, but during merging with your changes it got lost ...

File: cse/lang/NltkStopwordFilter.py (hunk 4 above)
Comment (SebastianSchmidl) [reply cse/lang/NltkStopwordFilter.py:16]: see above

File: cse/indexing/MainPostingListIndex.py (hunk 6 above)
Comment (SebastianSchmidl) [reply cse/indexing/MainPostingListIndex.py:68]: this is a copy paste issue
Should be changed!

File: cse/indexing/MainPostingListIndex.py (hunk 7 above)
Comment (SebastianSchmidl) [reply cse/indexing/MainPostingListIndex.py:79]: Yeah. You're right. Will change that soon.

Comment (CodeLionX) [reply cse/indexing/MainPostingListIndex.py]: Ok, thanks. Will change that

File: cse/indexing/createIndex.py (hunk 8 above)
Comment (SebastianSchmidl) [reply cse/indexing/createIndex.py:2]: Was renamed to `DocumentIndex`. I have created a list of ToDos ;)

File: cse/indexing/InvertedIndexReader.py (hunk 9 above)
Comment (SebastianSchmidl) [reply cse/indexing/InvertedIndexReader.py:28]: Python has method contracts for implementing container types, see https://docs.python.org/3/reference/datamodel.html?emulating-container-types#emulating-container-types
eg. I used: `__getitem__` or `__setitem__`

File: cse/indexing/createIndex.py (hunk 8 above)
Comment (Benedikt1992) [reply cse/indexing/createIndex.py:2]: `DocumentIndex' doesn't exist as well (at least in my copy of the repo)

Comment (SebastianSchmidl) [reply cse/indexing/createIndex.py:2]: sry typo: `DocumentMap` in `cse/lang/DocumentMap.py`
see refactoring in 01da17c

**Comment by SebastianSchmidl** [comment]
created PR #32 for addressing the discussed bugs

File: cse/CommentReader.py (hunk 2 above)
Comment (SebastianSchmidl) [reply cse/CommentReader.py:38]: No. I read all lines, but I just read the articleUrl and articleId once (second interation). It is/must be the same for all comments in the file. Additionally I skip the first row, it's the csv-header... we don't need that information.

**Pull request merged**